src/main/java/net/neoforged/neoforge/common/util/AttributeUtil.java

```diff
@@ -0,0 +1,356 @@
+                }
+
+                for (Operation op : Operation.values()) {
+                    int i = op.ordinal();
+
+                    // If the merged value comes out to 0, just ignore the whole stack
+                    if (sums[i] == 0) {
+                        continue;
+                    }
+
+                    // Handle merged modifier stacks by creating a "fake" merged modifier with the underlying value.
+                    if (merged[i]) {
+                        TextColor color = attr.value().getMergedStyle(sums[i] > 0);
+                        var fakeModif = new AttributeModifier(AttributeUtil.FAKE_MERGED_ID, sums[i], op);
+                        MutableComponent comp = attr.value().toComponent(fakeModif, ctx.flag());
+                        tooltip.accept(comp.withStyle(comp.getStyle().withColor(color)));
+                        if (ctx.flag().hasShiftDown() && merged[i]) {
+                            shiftExpands.get(Operation.BY_ID.apply(i)).forEach(modif -> tooltip.accept(AttributeUtil.list().append(attr.value().toComponent(modif, ctx.flag()))));
+                        }
+                    } else {
+                        var fakeModif = new AttributeModifier(AttributeUtil.FAKE_MERGED_ID, sums[i], op);
+                        tooltip.accept(attr.value().toComponent(fakeModif, ctx.flag()));
+                    }
+                }
+            } else {
+                for (AttributeModifier m : modifs) {
+                    if (m.amount() != 0) {
+                        tooltip.accept(attr.value().toComponent(m, ctx.flag()));
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Adds tooltip lines for the attribute modifiers contained in a {@link PotionContents}.
+     * 
+     * @param list     The list of attribute modifiers generated by calling {@link MobEffect#createModifiers} for each mob effect instance on the potion.
+     * @param tooltips The tooltip consumer to add lines to.
+     */
+    public static void addPotionTooltip(List<Pair<Holder<Attribute>, AttributeModifier>> list, Consumer<Component> tooltips) {
+        if (!list.isEmpty()) {
+            tooltips.accept(CommonComponents.EMPTY);
+            tooltips.accept(Component.translatable("potion.whenDrank").withStyle(ChatFormatting.DARK_PURPLE));
+
+            for (Pair<Holder<Attribute>, AttributeModifier> pair : list) {
+                tooltips.accept(pair.getFirst().value().toComponent(pair.getSecond(), AttributeUtil.getTooltipFlag()));
+            }
+        }
+    }
+
+    /**
+     * Creates a sorted {@link TreeMultimap} used to ensure a stable iteration order of item attribute modifiers.
+     */
+    public static Multimap<Holder<Attribute>, AttributeModifier> sortedMap() {
+        return TreeMultimap.create(Comparator.comparing(Holder::value, Comparator.comparing(BuiltInRegistries.ATTRIBUTE::getKey, ResourceLocation::compareTo)), ATTRIBUTE_MODIFIER_COMPARATOR);
```

Comment:
Do we want this to use `compareNamespaced` instead of `compareTo` to compare the resource locations? Otherwise we can remove the explicit `ResourceLocation::compareTo` parameter, as `ResourceLocation` already implements `Comparable`. Though at that point we might want to just consider doing a single comparator: `Comparator.comparing(holder -> BuiltInRegistries.ATTRIBUTE.getKey(holder.value())` rather than creating a second wrapping thing just to get the value.

Reply:
I think using ResourceLocation#compareTo is more correct here, because it will produce a (probably) alphabetically-sorted list of attributes by path, rather than invoking the namespace as part of the equation.

Reply:
Also looks like we might just be able to flatten the comparator to `Comparator.comparing(Holder::getKey)`